feat: Add option to use own NSURLSession for transport

dKasabwala opened the pull request:
…on configuration.
#skip-changelog
## :scroll: Description

- We are a current customer of sentry, and we use custom socks proxy on our end, we have no way to set up socks proxy for sentry services. we have added change to provide hook with URL Session. 

## :bulb: Support Socks proxy and private session configuration with custom SSL Certs.

- Solve Custom SSL pinning 
- Provide support Socks Proxy

## :green_heart: How did you test it?

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [x] Review from the native team if needed.
- [X] No breaking change or entry added to the changelog.
- [X] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps

brustolin commented on Sources/Sentry/SentryTransportFactory.m:
```diff
@@ -29,9 +29,18 @@ @implementation SentryTransportFactory
+     
+        session = options.urlSession;
+    } else {
+        session = [NSURLSession sessionWithConfiguration:configuration
+                                                              delegate:options.urlSessionDelegate
+                                                        delegateQueue:nil];
```
m: You could move lines 30, 31 in here.

dKasabwala replied:
Moved `ephemeralSessionConfiguration` inside else statement.

brustolin commented on Sources/Sentry/Public/SentryOptions.h:
```diff
@@ -8,6 +8,8 @@ NS_ASSUME_NONNULL_BEGIN
 NS_SWIFT_NAME(Options)
 @interface SentryOptions : NSObject
 
+
```
h: We need description of this new property like the others.

dKasabwala replied:
Updated description.

brustolin reviewed (commented):
This change is pretty simple and gives the user more power, which in my opinion make sense in here.

SentryOptions has a `- (BOOL)validateOptions:(NSDictionary<NSString *, id> *)options
       didFailWithError:(NSError *_Nullable *_Nullable)error` method that need to be updated to accommodate the new property.

We also need some tests to make sure the new property its being used properly.

dKasabwala commented:
- Added unit tests around new options for urlSession
- updated `- (BOOL)validateOptions:(NSDictionary<NSString *, id> *)options didFailWithError:(NSError *_Nullable *_Nullable)error`
- new property description been added as well.

philipphofmann commented on Sources/Sentry/SentryTransportFactory.m:
```diff
@@ -27,11 +27,19 @@ @implementation SentryTransportFactory
+
+    NSURLSession *session;
+
+    if (options.urlSession) {
```
`m`: Please add some tests to the `SentryTransportFactoryTests` for this new functionality.

dKasabwala replied:
Updated unit test to cover 100% coverage for this file.

dKasabwala commented:
@philipphofmann Please get this approved, All checks are complete, we do need these changes next release version.

philipphofmann reviewed (commented):
The unit tests in CI are failing because the unit test don't compile @dKasabwala. You can ignore the other failing CI jobs. They fail due to our CI setup.

dKasabwala commented:
Fix up compiler error

philipphofmann reviewed (approved):
LGTM
